frontend/javascripts/oxalis/model/sagas/isosurface_saga.ts

```diff
@@ -506,6 +515,66 @@ function* _refreshIsosurfaceWithMap(
  * Precomputed Meshes
  *
  */
+
+// Avoid redundant fetches of mesh files for the same layer by
+// storing Deferreds per layer lazily.
+const fetchDeferredsPerLayer: Record<string, Deferred<Array<APIMeshFile>, unknown>> = {};
+function* maybeFetchMeshFiles(action: MaybeFetchMeshFilesAction): Saga<void> {
+  const { segmentationLayer, dataset, mustRequest, autoActivate, callback } = action;
+
+  if (!segmentationLayer) {
+    callback([]);
+    return;
+  }
+
+  const layerName = segmentationLayer.name;
+
+  function* maybeActivateMeshFile(availableMeshFiles: APIMeshFile[]) {
+    const currentMeshFile = yield* select(
+      (state) => state.localSegmentationData[layerName].currentMeshFile,
+    );
+    if (!currentMeshFile && availableMeshFiles.length > 0 && autoActivate) {
+      yield* put(updateCurrentMeshFileAction(layerName, availableMeshFiles[0].meshFileName));
+    }
+  }
+
+  // If a deferred already exists, the one can be awaited (regardless of
+  // whether it's finished or not) and its content used to call the callback.
+  // If mustRequest was set to true, a new deferred will be created which
+  // replaces the old one (old references to the first Deferred will still
+  // work and will be resolved by the corresponding saga execution).
+  if (fetchDeferredsPerLayer[layerName] && !mustRequest) {
+    const meshes = yield* call(() => fetchDeferredsPerLayer[layerName].promise());
+    yield* maybeActivateMeshFile(meshes);
+    callback(meshes);
+    return;
+  }
+  const deferred = new Deferred<Array<APIMeshFile>, unknown>();
+  fetchDeferredsPerLayer[layerName] = deferred;
+
+  const files = yield* select((state) => state.localSegmentationData[layerName].availableMeshFiles);
+
+  // Only send new get request, if it hasn't happened before (files in store are null)
+  // else return the stored files (might be empty array). Or if we force a reload.
+  if (!files || mustRequest) {
+    const availableMeshFiles = yield* call(
+      getMeshfilesForDatasetLayer,
+      dataset.dataStore.url,
+      dataset,
+      getBaseSegmentationName(segmentationLayer),
+    );
+    yield* put(updateMeshFileListAction(layerName, availableMeshFiles));
+    deferred.resolve(availableMeshFiles);
```

Comment:
I think, the scenario you described cannot occur, since there will always be a deferred if the mesh files were already fetched. However, the code didn't really make use of this fact (and thus was hard to read). I simplified the code now accordingly :)

Reply:
Thank you, much more clear now :+1: